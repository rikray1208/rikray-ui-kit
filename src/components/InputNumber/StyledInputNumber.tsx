import styled, { css } from 'styled-components';

import { getColor, setFontSize, setShape } from '../../utils/helpers';

import type { InputNumberType } from './index';
import type { Shape, Size } from '../types';

function setPadding(size: Size) {
    switch (size) {
        case 'Large':
            return css`
                padding: ${({ theme }) => theme.InputNumber['padding-2'] + 'rem'};
            `;
        case 'Medium':
            return css`
                padding: ${({ theme }) => theme.InputNumber['padding-1.5'] + 'rem'};
            `;
        case 'Small':
            return css`
                padding: ${({ theme }) => theme.InputNumber['padding-1'] + 'rem'}
                    ${({ theme }) => theme.InputNumber['padding-2'] + 'rem'};
            `;
    }
}
export const InputWrapper = styled.div<{ variant: InputNumberType; size: Size; shape: Shape }>`
    ${({ theme, variant, size, shape }) => css`
        display: flex;
        width: min-content;
        align-items: center;

        background-color: ${theme.InputNumber.colorBgContainer};
        border-width: ${theme.InputNumber.borderWidth};
        border-style: ${theme.InputNumber.borderStyle};
        border-color: ${theme.InputNumber.colorBorderBase};
        outline: none;
        border-color: ${theme.InputNumber.colorBorderBase};
        ${setShape(shape, 'InputNumber')};

        &.focused,
        &:focus {
            box-shadow: 0 0 0 ${theme.Input.shadowWidthBase}px ${getColor(variant, 'InputNumber', { isShadow: true })};
            border-color: ${getColor(variant, 'InputNumber')};
        }

        &:hover {
            border-color: ${getColor(variant, 'InputNumber', { isHover: true })};
        }
        ${setFontSize(size, 'InputNumber', { less: size === 'Large' })}
        ${setPadding(size)}
    `}
`;

export const StyledInputNumber = styled.input<{ isControls: boolean }>`
    &::-webkit-outer-spin-button,
    &::-webkit-inner-spin-button {
        -webkit-appearance: none;
        margin: 0;
    }
    &[type='number'] {
        -moz-appearance: textfield;
    }

    padding: 0;

    border: none;
    outline: none;

    text-align: ${({ isControls }) => !isControls && 'center'};
    font: inherit;
`;

export const InputControls = styled.div<{ variant: InputNumberType }>`
    ${({ variant, theme }) => css`
        display: flex;
        flex-direction: column;
        justify-content: center;
        margin-left: auto;
        cursor: pointer;

        width: auto;

        > svg {
            color: ${theme.InputNumber.colorBorderBase};
        }

        > svg:hover {
            color: ${getColor(variant, 'InputNumber')};
        }
    `}
`;
