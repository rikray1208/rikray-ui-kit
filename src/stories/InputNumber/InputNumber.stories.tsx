import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import { InputNumber as InputNumberComp, ThemeProvider } from '../../index';

const decorators = [
    (Story: any) => (
        <ThemeProvider>
            <div>
                <Story />
            </div>
        </ThemeProvider>
    ),
];

export default {
    title: 'Design System/Data Entry/InputNumber',
    component: InputNumberComp,
    decorators: decorators,
} as ComponentMeta<typeof InputNumberComp>;

const Template: ComponentStory<typeof InputNumberComp> = (args) => <InputNumberComp {...args} />;

export const InputNumber = Template.bind({});
InputNumber.args = {
    controls: true,
    max: 10,
    min: 2,
    defaultValue: 2,
    step: 2,
    size: 'Medium',
};
