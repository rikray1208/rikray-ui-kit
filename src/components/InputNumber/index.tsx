import { ChevronDown, ChevronUp } from 'lucide-react';
import React, { FC, useEffect, useState } from 'react';

import { InputControls, InputWrapper, StyledInputNumber } from './StyledInputNumber';

import { Shape, Size, Status } from '../types';

export type InputNumberType = Status | 'primary';
export interface InputNumberProps {
    type?: InputNumberType;
    size?: Size;
    shape?: Shape;
    defaultValue?: number;
    max?: number;
    min?: number;
    step?: number;
    controls?: boolean;
    style?: React.CSSProperties;
    className?: string;
}
const InputNumber: FC<InputNumberProps> = (props) => {
    const {
        type = 'primary',
        min,
        max,
        defaultValue,
        step = 1,
        controls = true,
        size = 'Medium',
        shape = 'SemiRound',
        style,
        className = '',
    } = props;
    const [value, setValue] = useState<string | number>('');
    const [wrapperClassName, setWrapperClassName] = useState('InputNumberWrapper');
    function valueHandle(event: React.ChangeEvent<HTMLInputElement>) {
        if (event.target.value !== '') {
            const number = parseInt(event.target.value, 10);

            if (max && min) {
                if (number <= max && number >= min) {
                    setValue(number);
                }
            } else {
                if (max && number <= max) {
                    setValue(number);
                }

                if (min && number >= min) {
                    setValue(number);
                }
            }
        } else {
            setValue(event.target.value);
        }
    }
    function increment() {
        function changeValue(prev: string | number) {
            if (typeof prev === 'number') {
                if (max && max >= prev + step) {
                    return step + prev;
                } else {
                    return prev;
                }
            } else {
                return step;
            }
        }
        setValue(changeValue);
    }
    function decrement() {
        function changeValue(prev: string | number) {
            if (typeof prev === 'number') {
                if (typeof min === 'number' && min <= prev - step) {
                    return prev - step;
                } else {
                    return prev;
                }
            } else {
                return prev;
            }
        }

        setValue(changeValue);
    }
    function focusHandler() {
        setWrapperClassName('InputNumberWrapper focused');
    }

    function blurHandler() {
        setWrapperClassName('InputNumberWrapper');
    }

    useEffect(() => {
        if (defaultValue) {
            setValue(defaultValue);
        }
    }, [defaultValue]);

    return (
        <InputWrapper
            style={style}
            tabIndex={-1}
            variant={type}
            size={size}
            shape={shape}
            className={wrapperClassName + className}
        >
            <StyledInputNumber
                className='NumberInput'
                onFocus={focusHandler}
                onBlur={blurHandler}
                onChange={valueHandle}
                value={value}
                max={max}
                min={min}
                step={step}
                type={'number'}
                isControls={controls}
            />
            {controls && (
                <InputControls className='Controls' variant={type}>
                    <ChevronUp onClick={increment} size={16} />
                    <ChevronDown onClick={decrement} size={16} />
                </InputControls>
            )}
        </InputWrapper>
    );
};

export default InputNumber;
