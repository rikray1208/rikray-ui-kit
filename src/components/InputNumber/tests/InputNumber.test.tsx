import React from 'react';
import { create } from 'react-test-renderer';

import { ThemeProvider } from '../../../index';

import InputNumber from '../index';

describe('InputNumber', () => {
    it('should return a InputNumber', () => {
        const element = (
            <ThemeProvider>
                <InputNumber />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a InputNumber with default value', () => {
        const element = (
            <ThemeProvider>
                <InputNumber defaultValue={2} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a large InputNumber', () => {
        const element = (
            <ThemeProvider>
                <InputNumber size={'Large'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a small InputNumber', () => {
        const element = (
            <ThemeProvider>
                <InputNumber size={'Small'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a Rectangle InputNumber', () => {
        const element = (
            <ThemeProvider>
                <InputNumber shape={'Rectangle'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
    it('should return a Round InputNumber', () => {
        const element = (
            <ThemeProvider>
                <InputNumber shape={'Round'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
});
