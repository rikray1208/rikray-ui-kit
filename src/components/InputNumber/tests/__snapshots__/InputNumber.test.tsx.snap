// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputNumber should return a InputNumber 1`] = `
<div
  className="sc-iAEawV hAjudQ InputNumberWrapper"
  shape="SemiRound"
  size="Medium"
  tabIndex={1}
>
  <input
    className="sc-eeMvmM dfHLYm NumberInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    step={1}
    type="number"
    value=""
  />
  <div
    className="sc-cUEOzv kJxOtX Controls"
  >
    <svg
      className="lucide lucide-chevron-up"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="18 15 12 9 6 15"
      />
    </svg>
    <svg
      className="lucide lucide-chevron-down"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`InputNumber should return a InputNumber with default value 1`] = `
<div
  className="sc-iAEawV hAjudQ InputNumberWrapper"
  shape="SemiRound"
  size="Medium"
  tabIndex={1}
>
  <input
    className="sc-eeMvmM dfHLYm NumberInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    step={1}
    type="number"
    value=""
  />
  <div
    className="sc-cUEOzv kJxOtX Controls"
  >
    <svg
      className="lucide lucide-chevron-up"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="18 15 12 9 6 15"
      />
    </svg>
    <svg
      className="lucide lucide-chevron-down"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`InputNumber should return a Rectangle InputNumber 1`] = `
<div
  className="sc-iAEawV cSlQDg InputNumberWrapper"
  shape="Rectangle"
  size="Medium"
  tabIndex={1}
>
  <input
    className="sc-eeMvmM dfHLYm NumberInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    step={1}
    type="number"
    value=""
  />
  <div
    className="sc-cUEOzv kJxOtX Controls"
  >
    <svg
      className="lucide lucide-chevron-up"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="18 15 12 9 6 15"
      />
    </svg>
    <svg
      className="lucide lucide-chevron-down"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`InputNumber should return a Round InputNumber 1`] = `
<div
  className="sc-iAEawV gGoIHP InputNumberWrapper"
  shape="Round"
  size="Medium"
  tabIndex={1}
>
  <input
    className="sc-eeMvmM dfHLYm NumberInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    step={1}
    type="number"
    value=""
  />
  <div
    className="sc-cUEOzv kJxOtX Controls"
  >
    <svg
      className="lucide lucide-chevron-up"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="18 15 12 9 6 15"
      />
    </svg>
    <svg
      className="lucide lucide-chevron-down"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`InputNumber should return a large InputNumber 1`] = `
<div
  className="sc-iAEawV bDeEfn InputNumberWrapper"
  shape="SemiRound"
  size="Large"
  tabIndex={1}
>
  <input
    className="sc-eeMvmM dfHLYm NumberInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    step={1}
    type="number"
    value=""
  />
  <div
    className="sc-cUEOzv kJxOtX Controls"
  >
    <svg
      className="lucide lucide-chevron-up"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="18 15 12 9 6 15"
      />
    </svg>
    <svg
      className="lucide lucide-chevron-down"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`InputNumber should return a small InputNumber 1`] = `
<div
  className="sc-iAEawV UTcx InputNumberWrapper"
  shape="SemiRound"
  size="Small"
  tabIndex={1}
>
  <input
    className="sc-eeMvmM dfHLYm NumberInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    step={1}
    type="number"
    value=""
  />
  <div
    className="sc-cUEOzv kJxOtX Controls"
  >
    <svg
      className="lucide lucide-chevron-up"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="18 15 12 9 6 15"
      />
    </svg>
    <svg
      className="lucide lucide-chevron-down"
      fill="none"
      height={16}
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;
